### PR TITLE
ci: replace codecov with sonarcloud

### PR DESCRIPTION
**Pull request by crodriguezvega**
## Description

My GitHub workflow fu is not really strong, so I am doing a bit of guessing here. :D I would appreciate a look from CI masters @chatton or @DimitrisJim. :) 

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

## Summary by CodeRabbit

- **Chores**
	- Updated project documentation and links to refer to SonarCloud for test coverage.
- **New Features**
	- Integrated SonarCloud for enhanced code quality and security analysis across workflows.
- **Refactor**
	- Optimized Continuous Integration (CI) workflows for efficiency and added conditions for artifact uploads.
- **Documentation**
	- Updated development setup documentation with the new SonarCloud integration details.

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -43,4 +43,4 @@ write a little note why.
 - [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
 - [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
 - [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
-- [ ] Review `Codecov Report` in the comment section below once CI passes.
+- [ ] Review `SonarCloud Report` in the comment section below once CI passes.
```
[the remark this answers is not in the document]
Comment (chatton) [reply .github/PULL_REQUEST_TEMPLATE.md:46]: good bot!

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -104,53 +104,44 @@ jobs:
...
-            sed -i.bak "/$(echo $filename | sed 's/\//\\\//g')/d" coverage.txt
-          done
-        if: env.GIT_DIFF
-      - uses: codecov/codecov-action@v4
         with:
-          file: ./coverage.txt
-        if: env.GIT_DIFF
+          name: '${{ github.sha }}-03-coverage'
+      - name: sonarcloud
+        if: ${{ env.GIT_DIFF && !github.event.pull_request.draft }}
```
Comment (chatton) [inline .github/workflows/test.yml:143]: are we missing the null check for `secrets.SONAR_TOKEN` here?

Comment (chatton) [reply .github/workflows/test.yml:143]: thinking a bit more, we probably don't need that check at all as we're passing it in in the same workflow, we would just be checking that `secrets.SONAR_TOKEN` is set basically, maybe unnecessary.

File: modules/apps/callbacks/sonar-project.properties [hunk 3]
```diff
@@ -0,0 +1,14 @@
+sonar.projectKey=ibc-go-callbacks
```
Comment (chatton) [inline modules/apps/callbacks/sonar-project.properties:1]: I don't have any experience with this particular tool, so I'm happy to trust you with these configs and if there are any issues we can worry about it in a follow up :muscle:

**Review by chatton — APPROVED** [review]
Overall the workflow files looks good to me! The rabbit bot caught a few issues I think we should address (typo in path + choose a specific version of the workflow)

I'm happy to add these new workflows and address any issues we run into in a follow up. :+1:

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (chatton) [reply .github/PULL_REQUEST_TEMPLATE.md:46]: \_(ツ)_/¯

File: modules/apps/callbacks/sonar-project.properties (hunk 3 above)
Comment (crodriguezvega) [reply modules/apps/callbacks/sonar-project.properties:1]: I took the SDK as reference, so I hope it is correct. :)

**Review by damiannolan — APPROVED** [review]
Let's give it a go! :)

**Pull request merged**